state/bakerystorage/storage.go

```diff
@@ -34,70 +38,56 @@ type storageDoc struct {
 	ExpireAt time.Time `bson:"expire-at,omitempty"`
 }
 
-// ExpireAt implements ExpirableStorage.ExpireAt.
-func (s *storage) ExpireAt(expireAt time.Time) ExpirableStorage {
-	return &storage{config: s.config, expireAt: expireAt}
+type legacyRootKey struct {
+	RootKey []byte
 }
 
-// ExpireAfter implements ExpirableStorage.ExpireAt.
+// ExpireAfter implements ExpirableStorage.ExpireAfter.
 func (s *storage) ExpireAfter(expireAfter time.Duration) ExpirableStorage {
-	return &storage{config: s.config, expireAfter: expireAfter}
+	newStorage := *s
+	newStorage.expireAfter = expireAfter
+	return &newStorage
 }
 
-// Put implements bakery.Storage.Put.
-func (s *storage) Put(location, item string) error {
-	coll, closer := s.config.GetCollection()
+func (s *storage) RootKey() ([]byte, []byte, error) {
+	storage, closer := s.getStorage()
 	defer closer()
-
-	doc := storageDoc{
-		Location: location,
-		Item:     item,
-	}
-	expireAt := s.expireAt
-	if expireAt.IsZero() && s.expireAfter > 0 {
-		expireAt = s.config.Clock.Now().Add(s.expireAfter)
-	}
-	if !expireAt.IsZero() {
-		// NOTE(axw) we subtract one second from the expiry time, because
-		// the expireAfterSeconds index we create is 1 and not 0 due to
-		// a limitation in the mgo EnsureIndex API.
-		doc.ExpireAt = expireAt.Add(-1 * time.Second)
-	}
-	_, err := coll.Writeable().UpsertId(location, doc)
-	if err != nil {
-		return errors.Annotatef(err, "cannot store item for location %q", location)
-	}
-	return nil
+	return storage.RootKey()
 }
 
-// Get implements bakery.Storage.Get.
-func (s *storage) Get(location string) (string, error) {
+func (s *storage) getStorage() (bakery.Storage, func()) {
 	coll, closer := s.config.GetCollection()
-	defer closer()
+	return s.config.GetStorage(s.rootKeys, coll, s.expireAfter), closer
+}
 
-	var i storageDoc
-	err := coll.FindId(location).One(&i)
+func (s *storage) Get(id []byte) ([]byte, error) {
+	storage, closer := s.getStorage()
+	defer closer()
+	i, err := storage.Get(id)
 	if err != nil {
```

Comment:
We should check for NotFound and to a legacy get only in that case. If there's any other error we should return it.